### ISSUE TITLE
StructurePruner how out_mask affect the subnet module?

### ISSUE THREAD

**Issue by Haijunlv** [open-mmlab/mmrazor#160]
### Checklist
- I have searched related issues but cannot get the expected help.
- I have read related documents and don't know what to do.

### Describe the question you meet
I tried to read StructurePruner code. And found one question.
From my view, out_mask is used to masking the layer weight.
https://github.com/open-mmlab/mmrazor/blob/master/mmrazor/models/pruners/structure_pruning.py#L467
i can only find one  function of deploy_subnet()  to mask the weight. But it seems that supernet pre-training and search for subnet on the trained supernet code  do not use the deploy_subnet().   deploy_subnet() is used in subnet retraining stage
https://github.com/open-mmlab/mmrazor/blob/master/mmrazor/models/pruners/structure_pruning.py#L542
https://github.com/open-mmlab/mmrazor/blob/master/mmrazor/models/algorithms/autoslim.py#L88

in AutoSlim.train_step(), 
pruner.sample_subnet() and pruner.set_subnet(subnet_dict)  used to set a new subnet. But set_subnet() seems only set module.in_mask and module.out_mask.

i can find the affect of in_mask  at https://github.com/open-mmlab/mmrazor/blob/master/mmrazor/models/pruners/structure_pruning.py#L434.
But i can not find the affect of out_mask in supernet pretaining stage and search stage

[here]

### Post related information
1. The output of `pip list | grep "mmcv\|mmrazor\|^torch"`
[here]
2. Your config file if you modified it or created a new one.
```python
[here]
```
3. Your train log file if you meet the problem during training.
[here]
4. Other code you modified in the `mmrazor` folder.
[here]

**Comment by HIT-cwh** [comment]
Hi! Thank you for your issue. 
You are right, ``module.out_mask`` is useless during supernet training and searching. We just save ``module.out_mask`` to export subnet conveniently (ref to [here](https://github.com/open-mmlab/mmrazor/blob/2dad24044d7f1dad88f20221f8fc071dd40fdd4f/mmrazor/models/pruners/structure_pruning.py#L312-L331)).
The key and value of ``subnet_dict`` are ``space_id`` and the corresponding ``out_mask``, respectively. Note that the ``in_mask`` of a module is the ``out_mask`` of its parent module, typically. We use ``in_mask`` to mask the input rather than ``out_mask`` to mask the output in ``modify_conv_forward`` and ``modify_fc_forward``,  so that we only need to mask the input of the conv modules traced by ``trace_non_pass_path``. If we use ``out_mask`` to mask the output, we have to mask all modules between two conv modules.      
If you have any other questions, pls let me know.

**Comment by Haijunlv** [comment]
thanks for replying so quickly. my question is solved. thank you .
